### ISSUE TITLE
Expand router executable support to recognize all "single machine" queries

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#501]
@anarazel: (with minor edits) A customer runs queries to look at interesting "groups" of users, who then could be targeted.

This customer's current query generator, due to our limitations, essentially performs queries against a table residing on a single shard, retrieves the result, and then sends another query with huge identifier ``IN (...)`` or identifier <> ``ANY(...)`` lists.  To produce the desired end result this step happens repeatedly.

There's some other inefficiencies like useless joins, but they're basically just making bad things worse.

The only way I've found so far to speed up their queries substantially is to rewrite them as a series of LATERAL joins. That way the intermediary result sets don't have to be kept in memory, and many of the
later queries can be implemented as relatively efficient and small subquery scans.

Obviously we currently can't run the converted ``LATERAL`` queries with Citus. But as the data distribution guarantees that all the data involved in one query always resides on one worker, we have some options:

1. Provide them with an interface to query the worker and shard names for a specific project_key, and have them query that worker directly; replacing tablenames themselves.
2. Expand router executor's detection logic so it recognizes that all tables/shards involved in a placement reside on a single node, allowing much more complex queries to be pushed down.

I think (2) would help a number of other users too, besides obviously being less work for interstate. So I think that might be a viable path forward.

**Comment by ozgune** [comment]
@mtuncer: (with minor edits) This issue is related to #448. What I proposed in #448 won't fix the issues listed here, however implementing this level of support would fix #448.

We could extend #448 to support joins under certain conditions. Each table has a filter on partition column to reduce the query into a single shard, and join is on the partition column. We could probably relax this a bit once we know more.

There were potential design issues that we need to be addressed.

1. Router planner needs to know about shard placements. We can offload this check to physical planner and call a function from there to make sure shards are co-located on the same worker.

2. Executor needs to execute the selected task only, it is unlikely that there would be an alternative task placement to fall back if the first query fails to execute for any reason.

3. We might not be honoring task assignment policy. We will definitely need to write a new one for router executable join queries.

4. @onderkalaci mentioned that selected task assignment might not be on the shards first replica that interstate workers have cached. This may causes performance issues since worker memory might not be enough to cache more shards. 

I think we can implement this one after addressing those concerns.

However, I start to worry that we are using router executor for catch-all type option to support all kinds of queries that does not fit well in our distributed planner.  We might be introducing another maintenance heavy item. @marcocitus warned about this when we were about to introduce router planner for select queries.

**Comment by marcocitus** [comment]
A way to simplify this task would be to require that all subqueries have specific filters for each of their table rte's instead of doing join pruning.

The implication of not having join pruning would be that instead of:
FROM a JOIN b ON (a.x = b.x) WHERE a.x = 5 AND ...

Users would have to write:
FROM a JOIN b ON (a.x = b.x) WHERE a.x = 5 AND b.x = 5 AND ...

Queries are safe to push down if they meet the following conditions:
- subqueries have a <distribution column> = <constant> filter for all their table rte's
- the distribution column constants are equal for all subqueries
- all tables are co-located

Even if the user writes complex queries that join with generated values or on something other than the distribution column, these queries would be safe, since the filter ensures that no information outside of the single shard (or co-located shards) is considered. 

We should be a little careful about co-location. If there are co-partitioned placements that are not on the same node, the query cannot be pushed down for some distribution column values. In that case, it might be better to error out on all queries to catch it early.

**Comment by ozgune** [comment]
@marcocitus / @anarazel -- we decided to start with a prototype on this issue to come up with a better understanding of the feature's design, scope, and timeline estimates.

**Comment by anarazel** [comment]
My current idea about how to implement this is, on a high level, the
following:

* during postgres planning collect all relevant
  RelOptInfo->baserestrictinfo's - these will contain information about
  the restrictions on invididual tables. If the condition / join-tree
  permit, it'll even distribute restrictions across joins.

* Use a RangeTable walker, go over all relations, check if the
  collected RestrictInfo leads to pruning down to a single shard. If
  not, abort. If so, build bitmap about workers with healthy placements

* Compute intersection between all bitmapsets. If there's one worker
  containing all required placement, execute queries there.

I think it'd be generally rather beneficial if we'd start using the
restrictinfos, not just in router, but also the normal planner. Afaics
there's a good number of scenarios where that'd computes more
restrictions than the ones citus' planner can come up with.


The difficulty with that plan is that before 9.5 `set_rel_pathlist_hook`
does not exist. That hook would allow us to efficiently collect the
RestrictInfos.  It looks to me that for 9.4 we'd likely have to redo a
lot of the hangups in planner.c/initsplan.c to get PlannerInfo in a
state after distribute_qual_to_rels() having been called.


@marcocitus  Does that plan roughly sound sensible to you?


My current inclination is to go ahead and implement the prototype
requiring 9.5 support. That'd allow us to estimate the cost of 9.4
support more accurately.

**Comment by mtuncer** [comment]
Here is the task list in order of implemantation
- [x] Refactor multi planner to directly create router plan, fall back to normal planner upon failure
- [ ] Pass original parse query to router planner in addition to modified one by standard_planner. Refactor router planner to remove old hacks (make_ands_explicit ...) to work around standart_planner modifications on the query.
- [ ] Set planner hooks to collect restriction information
- [ ] Perform shard pruning based on collected restrictions
- [ ] Check shard placement to see if all target shards are at the same worker
- [ ] Create router query and task

Each step contains making a router plannable decision. If it turns out that query is not router plannable, it would give up as early as possible.

**Comment by marcocitus** [comment]
> Use a RangeTable walker, go over all relations, check if the collected RestrictInfo leads to pruning down to a single shard. If not, abort. If so, build bitmap about workers with healthy placements

I'm still a bit worried about the user trying out queries like SELECT * FROM table WHERE x IN (3,4) where x is the partition key. It might work when you first try it, but when you put it into production you might suddenly find there are cases in which the query hits multiple shards and aborts. What if we apply the same restriction as we do currently (require clauses of the form <partition column> = <constant> AND ...)?

**Comment by anarazel** [comment]
On 2016-06-21 11:59:45 -0700, Marco Slot wrote:
> > Use a RangeTable walker, go over all relations, check if the collected RestrictInfo leads to pruning down to a single shard. If not, abort. If so, build bitmap about workers with healthy placements
> 
> I'm still a bit worried about the user trying out queries like SELECT
> * FROM table WHERE x IN (3,4) where x is the partition key. It might
> work when you first try it, but when you put it into production you
> might suddenly find there are cases in which the query hits multiple
> shards and aborts. What if we apply the same restriction as we do
> currently (require clauses of the form <partition column> = <constant>
> AND ...)?

I think that'd be a significant regression about which cases we can
apply this facility to. E.g. with intentionally 1 shard distributed
relations etc., it'd remove a fair bit of capability.

This isn't really an issue unique to router, it also exists with
co-located / broadcast joins and such.

**Comment by mtuncer** [comment]
Current Issues

- [ ] following query crashes in router executor, it used to be not router plannable, but  became router plannable since there is only one shard in the regression tests. Assertion failure is related to ```WITH (OIDS)``` option.
``` sql
CREATE TEMP TABLE customer_few (customer_key) WITH (OIDS) ON COMMIT DROP AS 
    (SELECT * FROM customer WHERE c_nationkey = 1 ORDER BY c_custkey LIMIT 10);
```
- [ ] if the query turned out to be not router plannable after shard pruning in router planner, shard pruning is performed again in physical planner. Perhaps it should be done only once.

- [ ] parameterized prepared statements does not work

- [ ] outer join regression tests needs to be revisited since those queries are now router plannable. They don't go through outer join logic.

- [ ] multi_join_order_additional test needs to be revisited

- [ ] router executor displays results for zero-shard select queries whereas real-time executor does not.

**Comment by ozgune** [comment]
@mtuncer -- quick side question. Once we merge in these changes, will we cover all scenarios mentioned in #448?

**Comment by mtuncer** [comment]
Most of them are addressed, except recursive queries, and queries inside target list.

Subqueries, joins, unions, and CTEs work.

**Comment by ozgune** [comment]
@mtuncer -- will window functions work?

**Comment by mtuncer** [comment]
@ozgune They already work.
